nexus/db-queries/src/db/on_conflict_ext.rs

```diff
@@ -0,0 +1,36 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use diesel::dsl::sql;
+use diesel::expression::SqlLiteral;
+use diesel::prelude::*;
+use diesel::sql_types::Bool;
+use diesel::upsert::IncompleteOnConflict;
+
+/// An extension trait for incomplete [`ON CONFLICT`] expressions.
+pub trait IncompleteOnConflictExt {
+    type AsPartialIndexOutput;
+
+    // TODO: explain what this does.
+    fn as_partial_index(self) -> Self::AsPartialIndexOutput;
+}
+
+impl<Stmt, Target> IncompleteOnConflictExt
+    for IncompleteOnConflict<Stmt, Target>
+where
+    Target: DecoratableTarget<SqlLiteral<Bool>>,
+{
+    type AsPartialIndexOutput =
+        IncompleteOnConflict<Stmt, DecoratedAsPartialIndex<Target>>;
+
+    fn as_partial_index(self) -> Self::AsPartialIndexOutput {
+        #[allow(clippy::disallowed_methods)]
+        {
+            self.filter_target(sql::<Bool>("false"))
```

Comment:
So I do understand (somewhat) based on the discussion yesterday that "contradiction implies YOLO", so `false` **implies** all partial indexes.

Is there ever a valid reason to select a subset of partial indices? Would this be desirable where someone wants conflict on one index, but not another?

Reply:
> So I do understand (somewhat) based on the discussion yesterday that "contradiction implies YOLO", so false implies all partial indexes.

Specifically, `false` implies all the _constraints_ on the partial indexes for the given columns. For example, in this case we have a partial index:

```sql
CREATE UNIQUE INDEX IF NOT EXISTS lookup_silo_user_by_silo ON omicron.public.silo_user (
    silo_id,
    external_id
) WHERE
    time_deleted IS NULL;
```

If you specify `silo_id` and `external_id`, and pass in `WHERE false`, the database will be able to infer that you mean the constraint `time_deleted IS NULL`. If you specify a set of columns that doesn't have any kind of unique constraint on it, the query will error out.

> Is there ever a valid reason to select a subset of partial indices? Would this be desirable where someone wants conflict on one index, but not another?

From my testing against crdb, the set of columns has to match one or more `UNIQUE INDEX`es exactly. If there are multiple `UNIQUE INDEX` instances on a set of columns, then `WHERE false` will match all of them.

I think there are theoretical cases where it's possible that being able to specify some narrower constraint is useful, but I can't imagine we'll be needing it in omicron any time soon.